staging/src/k8s.io/apimachinery/pkg/labels/selector.go

```diff
@@ -166,7 +166,6 @@ func NewRequirement(key string, op selection.Operator, vals []string) (*Requirem
 			return nil, err
 		}
 	}
-	sort.Strings(vals)
```

Comment:
I'm not sure which option is better: if sort in `NewRequirement`, it only sort once; if sort in `String()`, it'll sort it every time when call `String()`.

According to the description at #35194, it improve performance as follow:

> Early results:
>
> * scheduler throughput it ~15% higher
> * apiserver cpu-usage is also lower (seems to be also ~10-15%)

IMO, avoiding sort in `String()` by removing `set.Strings.List()` should contribute :)

Reply:
Would you try benchmark of scheduler? I'm not sure apiserver, but I think there should be some tools there.

Reply:
I tried benchmark test of scheduler, but didn't find any significant difference between the two solutions. And I realized that it might be current benchmark testcases doesn't cover the scenario we try to fix - multiple-value requirements. So I added a case like this: ([]string{"yellow", "green"} are designed in particular to hit sort())

```diff
diff --git a/test/integration/scheduler_perf/scheduler_bench_test.go b/test/integration/scheduler_perf/scheduler_bench_test.go
index fb28da4ec6..6ab25056d2 100644
--- a/test/integration/scheduler_perf/scheduler_bench_test.go
+++ b/test/integration/scheduler_perf/scheduler_bench_test.go
@@ -71,7 +71,27 @@ func BenchmarkSchedulingAntiAffinity(b *testing.B) {
 			benchmarkScheduling(test.nodes, test.existingPods, test.minPods, setupStrategy, testStrategy, b)
 		})
 	}
+}
 
+// BenchmarkSchedulingNodeAffinity benchmarks the scheduling rate of pods with
+// NodeAffinity rules when the cluster has various quantities of nodes and
+// scheduled pods.
+func BenchmarkSchedulingNodeAffinity(b *testing.B) {
+	tests := []struct{ nodes, existingPods, minPods int }{
+		{nodes: 500, existingPods: 250, minPods: 250},
+		{nodes: 500, existingPods: 5000, minPods: 250},
+	}
+	// The setup strategy creates pods with no affinity rules.
+	setupStrategy := testutils.NewSimpleWithControllerCreatePodStrategy("setup")
+	// The test strategy creates pods with node-affinity for each other.
+	testBasePod := makeBasePodWithNodeAffinity("color", []string{"yellow", "green"})
+	testStrategy := testutils.NewCustomCreatePodStrategy(testBasePod)
+	for _, test := range tests {
+		name := fmt.Sprintf("%vNodes/%vPods", test.nodes, test.existingPods)
+		b.Run(name, func(b *testing.B) {
+			benchmarkScheduling(test.nodes, test.existingPods, test.minPods, setupStrategy, testStrategy, b)
+		})
+	}
 }
 
 // makeBasePodWithAntiAffinity creates a Pod object to be used as a template.
@@ -99,6 +119,35 @@ func makeBasePodWithAntiAffinity(podLabels, affinityLabels map[string]string) *v
 	return basePod
 }
 
+// makeBasePodWithNodeAffinity creates a Pod object to be used as a template.
+// The Pod has a NodeAffinity requirement against nodes with the given expressions.
+func makeBasePodWithNodeAffinity(key string, vals []string) *v1.Pod {
+	basePod := &v1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			GenerateName: "node-affinity-",
+		},
+		Spec: testutils.MakePodSpec(),
+	}
+	basePod.Spec.Affinity = &v1.Affinity{
+		NodeAffinity: &v1.NodeAffinity{
+			RequiredDuringSchedulingIgnoredDuringExecution: &v1.NodeSelector{
+				NodeSelectorTerms: []v1.NodeSelectorTerm{
+					{
+						MatchExpressions: []v1.NodeSelectorRequirement{
+							{
+								Key:      key,
+								Operator: v1.NodeSelectorOpNotIn,
+								Values:   vals,
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+	return basePod
+}
+
 // benchmarkScheduling benchmarks scheduling rate with specific number of nodes
 // and specific number of pods already scheduled.
 // This will schedule numExistingPods pods before the benchmark starts, and at
```

Then I rerun it (the command is `RUN_BENCHMARK=true ./test-performance.sh`) a few times, still no big difference on the benchmark test result. They're both sth. like:

```
BenchmarkScheduling/100Nodes/0Pods-8                 200           7131056 ns/op
BenchmarkScheduling/100Nodes/1000Pods-8              100          12868582 ns/op
BenchmarkScheduling/1000Nodes/0Pods-8                100          56496372 ns/op
BenchmarkScheduling/1000Nodes/1000Pods-8             100          52169635 ns/op
BenchmarkSchedulingAntiAffinity/500Nodes/250Pods-8                   250          16827985 ns/op
BenchmarkSchedulingAntiAffinity/500Nodes/5000Pods-8                  250          25059125 ns/op
BenchmarkSchedulingNodeAffinity/500Nodes/250Pods-8                   250          13460290 ns/op
BenchmarkSchedulingNodeAffinity/500Nodes/5000Pods-8                  250          21225831 ns/op

0s      rate: 621       total: 621 (qps frequency: map[621:1])
1s      rate: 753       total: 1374 (qps frequency: map[621:1 753:1])
2s      rate: 776       total: 2150 (qps frequency: map[753:1 776:1 621:1])
3s      rate: 790       total: 2940 (qps frequency: map[753:1 776:1 790:1 621:1])
Scheduled 3000 Pods in 4 seconds (750 per second on average). min QPS was 621
```

Then I did another test to print a message in both solutions (right before `sort.Strings()`). The result shows that:

- for the solution to put safeSort() in NewReuirement, the message shows up `30993 ` times
- for the solution to put safeSort() in String(), the message shows up `0` times

So it shows although the first solution runs sorting for 30k times, the benchmark didn't show major differences - I guess it's b/c we're justing sorting a two elements slice, that's not a big deal - but it's a common case, I believe nobody would use an affinity having a big values set as "matchExpressions".

Reply:
if so, both are ok to me :)

Reply:
> I believe nobody would use an affinity having a big values set as "matchExpressions".

I'm not sure what you think is big, but in the motivating #66298, [my case has nine elements today](https://github.com/kubernetes/kubernetes/issues/66298#issuecomment-405651546), and will likely accumulate more in the near future.

Reply:
@seh I can't tell an exact number without running benchmark tests. But sorting dozens of elements would be trivial. And also note that in current solution (safeSort() in String() of Requirement), it doesn't hit String() - internal objects are kept what it is - so there is no sorting.